File: RAO.sh/Microsoft/Day IV/2 Number Of Ways To Arrive At Destination/NumberOfWaysToArriveAtDestination.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
#define fastIO ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
#define endl "\n" 
// #define MOD 1000000007

class Solution{
    int MOD = 1e9 + 7;
public:
    int countPaths(int n, vector<vector<int>>& roads){
        vector<pair<int, int>> ADJ[n];
        for(auto road : roads){
            ADJ[road[0]].push_back({road[1], road[2]});
            ADJ[road[1]].push_back({road[0], road[2]});
        }

        // Priority Queue for Dijkstra's algo
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> PQ;
        PQ.push({0, 0});    // Pushing the source node

        // Distance and Ways array
        vector<int> dist(n, INT_MAX), ways(n, 0);
        dist[0] = 0;
        ways[0] = 1;

        // Dijkstra
        while(!PQ.empty()){
            int distance = PQ.top().first, node = PQ.top().second;
            PQ.pop();

            for(auto child : ADJ[node]){
                int adjNode = child.first;
                int adjDist = child.second;

                // Arriving for the first time
                if(distance + adjDist < dist[adjNode]){
                    dist[adjNode] = distance + adjDist;
                    ways[adjNode] = ways[node];

                    PQ.push({dist[adjNode], adjNode});
                }
                // Arriving for the not first time with same distance
                else if(distance + adjDist == dist[adjNode]){
                    ways[adjNode] = (ways[adjNode] +  ways[node])%MOD;
                }
            }
        }

        return ways[n-1];
    }
};

void solve(){
    
    
    return;
}

int main(){
    fastIO
    ll t;
    cin>>t;
           
    while(t--){
        solve();
    }
              
    return 0;
}

File: Ravi/Microsoft/Day 4/NumberOfWaysToArriveAtDestination.java
class Pair {
    long f;
    int s;
    
    Pair(long a, int b) {
        f = a;
        s = b;
    }
}

class Solution {
    public int countPaths(int n, int[][] roads) {
        
        ArrayList<Pair> adj[] = new ArrayList[n];
        for(int i = 0; i < n; i++) adj[i] = new ArrayList<>();
        
        for(int[] r : roads) {
            adj[r[0]].add(new Pair(r[1], r[2]));
            adj[r[1]].add(new Pair(r[0], r[2]));
        }
        
        long[] dis = new long[n];
        int[] times = new int[n];
        
        for(int i = 0; i < n; i++) dis[i] = (long)(1e18);
        
        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> a.f < b.f ? -1 : 1);
        pq.add(new Pair(0l, 0));
        times[0] = 1;
        dis[0] = 0;
        
        while(!pq.isEmpty()) {
            Pair p = pq.poll();
            int node = p.s;
            long w = p.f;
            
            for(Pair i : adj[node]) {
                if(dis[(int)i.f] > (long)(i.s + w)) {
                    dis[(int)i.f] = i.s + w;
                    pq.add(new Pair(dis[(int)i.f], (int)i.f));
                    times[(int)i.f] = times[node];
                } else if(dis[(int)i.f] == (long)(i.s + w)) {
                    times[(int)i.f] = (times[(int)i.f] + times[node]) % ((int)(1e9) + 7);
                }
            }
            
        }
        
        return times[n - 1];
    }
}
